label_studio_ml/examples/grounding_dino/dino.py

```diff
@@ -0,0 +1,212 @@
+    from mobile_sam import SamPredictor, sam_model_registry
+
+    model_checkpoint = MOBILESAM_CHECKPOINT
+    reg_key = 'vit_t'
+elif USE_SAM:
+    from segment_anything import SamPredictor, sam_model_registry
+
+    model_checkpoint = SAM_CHECKPOINT
+    reg_key = 'vit_h'
+
+
+DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
+
+
+class DINOBackend(LabelStudioMLBase):
+
+    def __init__(self, project_id, **kwargs):
+        super(DINOBackend, self).__init__(**kwargs)
+
+        self.label = None
+
+        self.from_name, self.to_name, self.value = None, None, None
+
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+
+        if USE_MOBILE_SAM or USE_SAM:
+            sam = sam_model_registry[reg_key](checkpoint=model_checkpoint)
+            sam.to(device=self.device)
+            self.predictor = SamPredictor(sam)    
+            
+
+    def predict(self, tasks: List[Dict], context: Optional[Dict] = None, **kwargs) -> List[Dict]:
+
+        self.from_name, self.to_name, self.value = self.get_first_tag_occurence('RectangleLabels', 'Image')
+
+        TEXT_PROMPT = context['result'][0]['value']['text'][0]
+
+        self.label = TEXT_PROMPT.strip("_SAM") # make sure that using as text prompt allows you to label it a certain way
+
+        
+        all_points = []
+        all_scores = []
+        all_lengths = []
+
+        for task in tasks:
+
+            raw_img_path = task['data']['image']
+
+            try:
+                img_path = get_image_local_path(
+                    raw_img_path,
+                    label_studio_access_token=LABEL_STUDIO_ACCESS_TOKEN,
+                    label_studio_host=LABEL_STUDIO_HOST
+                )
+            except:
+                img_path = raw_img_path
+
+            src, img = load_image(img_path)
+
+            boxes, logits, phrases = predict(
+                model=groundingdino_model,
+                image=img,
+                caption=TEXT_PROMPT.strip("_SAM"),
+                box_threshold=BOX_TRESHOLD,
+                text_threshold=TEXT_TRESHOLD,
+                device=DEVICE
+            )
+
+            H, W, _ = src.shape
+
+            boxes_xyxy = box_ops.box_cxcywh_to_xyxy(boxes) * torch.Tensor([W, H, W, H])
+
+            points = boxes_xyxy.cpu().numpy()
+
+            for point, logit in zip(points, logits):
+                all_points.append(point)
+                all_scores.append(logit)
+                all_lengths.append((H, W))
+
+        if USE_MOBILE_SAM or USE_SAM:
+            predictions = self.get_sam_results(img_path, all_points, all_lengths)
+        else:
+            predictions = self.get_results(all_points, all_scores, all_lengths)
+        
+        return predictions
+
+    def get_results(self, all_points, all_scores, all_lengths):
+        
+        results = []
+        
+        for points, scores, lengths in zip(all_points, all_scores, all_lengths):
+            # random ID
+            label_id = str(uuid4())[:4]
```

Comment:
Why not take the whole uuid, instead of truncating it?

Reply:
Not sure as this line was used from @niklub's SAM implementation which ended up fixing disappearing annotations

Reply:
@pakelley we may better check this with @hlomzik but to the best of my knowledge, a per-list unique `id` is a requirement for interactive predictions. Truncating it only for the sake of the payload size

Reply:
Gotcha yeah, my concern is mainly on how unique we can expect 4 characters to be. Do these label ids need to be unique only within their project, or across the entire database?
Honestly, either way 4 characters seems pretty low. uuid4 output is hexadecimal, so that gives us 16^4 or 65,536 possibilities. Guaranteed conflict above that, but likely conflict after... a few hundred? (rough birthday problem logic)

Reply:
I'd recommend we use at least up to the first "-" if this only has to be unique within the project (so 8 digits, ~400M possibilities, likely collision not until over 1M)
If it has to be global, I'd recommend up to the second "-" (12 digits, not including the first "-", 300T possibilities, likely collision not until many millions of records)
All that said, maybe I'm just missing that there really aren't that many ids being generated here?

Reply:
ids are unique per task (annotation in our terms), so don't worry about the whole database =)

but possibilities spread over the number of tasks, so if you have 4 digits (1/65k chance), have only 10 regions in every task, but have 100k of tasks, you'll definitelly get collision somewhere.

Reply:
Ok, I'll change this to truncate at 8 digits